libcloud/compute/drivers/profitbricks.py

```
@@ -0,0 +1,1569 @@
+            extra={
+                'datacenter_id': datacenter_id,
+                'datacenter_version': datacenter_version,
+                'server_id': server_id,
+                'lan_id': lan_id,
+                'internet_access': internet_access,
+                'mac_address': mac_address,
+                'dhcp_active': dhcp_active,
+                'gateway_ip': gateway_ip,
+                'ips': ips})
+
+    def _to_location(self, data):
+
+        return NodeLocation(id=data["region"],
+                            name=data["region"],
+                            country=data["country"],
+                            driver=self.connection.driver)
+
+    def _to_node_size(self, data):
+        """
+        Convert the PROFIT_BRICKS_GENERIC_SIZES into NodeSize
+        """
+        return NodeSize(id=data["id"],
+                        name=data["name"],
+                        ram=data["ram"],
+                        disk=data["disk"],
+                        bandwidth=None,
+                        price=None,
+                        driver=self.connection.driver,
+                        extra={
+                            'cores': data["cores"]})
+
+    def _wait_for_datacenter_state(
+            self,
+            datacenter,
+            state=PROVISIONING_STATE.get(NodeState.RUNNING)):
+        """
+        Private function that waits the datacenter
+        """
+        dc_operation_status = self.ex_describe_datacenter(datacenter[0])
+
+        timeout = 60 * 5
+        waittime = 0
+        interval = 5
+
+        while ((dc_operation_status[0].extra['provisioning_state']) ==
+                (self.PROVISIONING_STATE.get(NodeState.PENDING))) and (
+                waittime < timeout):
+            dc_operation_status = self.ex_describe_datacenter(datacenter[0])
+            if dc_operation_status[0].extra['provisioning_state'] == state:
+                break
+
+            waittime += interval
+            time.sleep(interval)
+
```

This method doesn't actually do anything different if a timeout occurs (`None` is implicitly returned in both cases).

On success, we should probably return object in question (latest reference to the `datacenter` object in this case) and on time out, we should throw a timeout exception.

Same goes for `_wait_for_storage_volume_state`.